core/model/src/main/java/org/eclipse/rdf4j/model/impl/SimpleLiteral.java

```diff
@@ -130,11 +145,26 @@ protected void setDatatype(IRI datatype) {
 		this.datatype = datatype;
 	}
 
+	protected void setDatatype(XmlDatatypeEnum datatype) {
+		this.datatype = datatype.getIri();
+		this.xmlDatatypeCached = true;
+		this.xmlDatatype = datatype;
+	}
+
 	@Override
 	public IRI getDatatype() {
 		return datatype;
 	}
 
+	@Override
+	public XmlDatatypeEnum getXmlDatatypeEnum() {
+		if (!xmlDatatypeCached) {
```

Comment:
Is this separate boolean flag necessary? You could just do this with a simple null check, right?

Reply:
removed now that we've moved to optionals